modules/spring-config/src/main/java/org/mule/runtime/config/api/dsl/model/metadata/DslElementBasedValueProviderCacheIdGenerator.java

```diff
@@ -0,0 +1,223 @@
+                                                                  ValueProviderModel valueProviderModel,
+                                                                  Map<String, ParameterModelInformation> parameterModelsInformation) {
+    List<ValueProviderCacheId> parts = new LinkedList<>();
+
+    parts.add(resolveValueProviderId(valueProviderModel));
+    parts.addAll(resolveActingParameterIds(valueProviderModel, parameterModelsInformation));
+    parts.addAll(resolveIdForInjectedElements(containerComponent, valueProviderModel));
+
+    return of(aValueProviderCacheId(fromElementWithName(resolveDslTag(containerComponent)
+        .orElse(getSourceElementName(containerComponent))).containing(parts)));
+  }
+
+  private List<ValueProviderCacheId> resolveIdForInjectedElements(DslElementModel<?> containerComponent,
+                                                                  ValueProviderModel valueProviderModel) {
+    if (!valueProviderModel.requiresConfiguration() && !valueProviderModel.requiresConnection()) {
+      return emptyList();
+    }
+    List<ValueProviderCacheId> injectableIds = new LinkedList<>();
+    Optional<DslElementModel<?>> configDslElementModel = resolveConfigName(containerComponent)
+        .flatMap(config -> locator.get(Location.builder().globalName(config).build()));
+    if (configDslElementModel.isPresent() && configDslElementModel.get().getModel() instanceof ConfigurationModel) {
+      if (valueProviderModel.requiresConfiguration()) {
+        resolveIdForInjectedElement(configDslElementModel.get())
+            .ifPresent(id -> injectableIds.add(aValueProviderCacheId(fromElementWithName("config: ").containing(id))));
+      }
+      if (valueProviderModel.requiresConnection()) {
+        configDslElementModel.get().getContainedElements().stream()
+            .filter(nested -> nested.getModel() instanceof ConnectionProviderModel).forEach(
+                                                                                            connectionProvider -> resolveIdForInjectedElement(connectionProvider)
+                                                                                                .ifPresent(id -> injectableIds
+                                                                                                    .add(aValueProviderCacheId(fromElementWithName("connection: ")
+                                                                                                        .containing(id)))));
+      }
+    }
+    return injectableIds;
+  }
+
+
+  private Optional<ValueProviderCacheId> resolveIdForInjectedElement(DslElementModel<?> injectedElement) {
+    if (!(injectedElement.getModel() instanceof EnrichableModel)) {
+      return empty();
+    }
+    List<ValueProviderCacheId> parts = new LinkedList<>();
+    List<String> parametersRequiredForMetadata =
+        ((EnrichableModel) injectedElement.getModel())
+            .getModelProperty(RequiredForMetadataModelProperty.class)
```

Comment:
RequiredForMetadataModelProperty.class, would this be fine?
Are we going to assume that also parameters required from Metadata are required for ValueProviders. Is this the case of connection/configuration elements only?

Reply:
I also thought about this and i think it's the best approach. 
First, we need to consider that this is only for when connections or configurations are being referenced from an operation or source valueProvider. Otherwise, when we are retrieving values for a Config, we don't use this annotation but the acting parameters provided by the ValueProviderModel.

So, when an operation's value provider makes use of a configuration for example, the parameters that define the values being retrieved should be the ones that define how to access the external system. We already know that for any MetadataResolver (Keys, Output, Input or Attributes) this is done by parameters annotated with @RequiredForMetadata (Or all of them by default). In this case, it makes sense to do the same. Otherwise, in which cases does a ValueProvider from an Operation or Source need a Config or Connection?